pkgs/fc/ceph/src/fc/ceph/maintenance/images_nautilus.py

```diff
@@ -273,75 +232,91 @@ def store_in_ceph(self, img):
+        td, filename = self.download_image()
         uncompressed = filename[0 : filename.rfind(".lz4")]
         subprocess.check_call(["unlz4", "-q", filename, uncompressed])
         os.unlink(filename)
         self.store_in_ceph(uncompressed)
-        logger.info("\tCreating snapshot %s", name)
-        run.rbd("snap", "create", self.volume + "@" + name)
-        run.rbd("snap", "protect", self.volume + "@" + name)
-
-    def flatten(self):
-        """Decouple VMs created from their base snapshots."""
-        logger.debug("Flattening child images for %s", self.release)
-        for snap in run.json.rbd("snap", "ls", self.volume):
-            for child in run.json.rbd(
-                "children", self.volume + "@" + snap["name"]
-            ):
+        logger.info("\tCreating snapshot %s", self.name)
+        run.rbd("snap", "create", self.snap_spec)
+        run.rbd("snap", "protect", self.snap_spec)
+
+    @property
+    def name(self):
+        return f"{self.release_name}-{self.image_hash_sha256}"
+
+    @property
+    def snap_spec(self):
+        return self.volume + "@" + self.name
+
+    def download_image(self):
+        """Fetches compressed image from Hydra.
+
+        Returns tmpdir handle and filename.
+        """
+        logging.debug("\t\tGetting %s", self.image_url)
+        r = requests.get(self.image_url, stream=True)
+        r.raise_for_status()
+        chksum = hashlib.sha256()
+        size = 0
+        td = tempfile.TemporaryDirectory(prefix="image.")
+        outfile = p.join(td.name, f"{self.release_name}-{self.envname}")
+        logging.debug("\t\tSaving to %s", outfile)
+        with open(outfile, "wb") as out:
+            for chunk in r.iter_content(4 * 2**20):
+                out.write(chunk)
+                chksum.update(chunk)
+                size += len(chunk)
+        if self.image_hash_sha256 != chksum.hexdigest():
+            raise RuntimeError(
+                f"Image checksum mismatch: expect={self.image_hash_sha256}, got={chksum.hexdigest()}"
+            )
+        return td, outfile
+
+    def cleanup(self) -> None:
+        """Cleanup of old, unneeded snaphshots.
+
+        As a precondition, VMs created from old base snapshots need to be
+        flattened (decoupled from that base). Then unprotecting and removing
+        is possible.
+
+        This can fail due to a race condition: flatten+unprotect is not an
+        atomic operation. Snapshots can still be in use after flattening due to
+        a new VM is cloned from the old snapshot in the meantime, hence
+        unprotecting fails.
+        We accept that race condition, just continue, and try again at the next
+        command run.
+        """
+        snaps: list = run.json.rbd("snap", "ls", self.volume)
+        snap_specs = [self.volume + "@" + snap["name"] for snap in snaps]
+
+        # hard invariant: always keep at least 1 snapshot per image
+        if len(snap_specs) == 1:
```

Comment:
Adopted the proposed changes with a slight change: I accept an empty list of snaps here, so that the assertion about a missing current snapshot is still triggered later in that case. This is a case we should still be notified about.